pdns/pdns_recursor.cc

```diff
@@ -136,6 +137,8 @@ thread_local std::unique_ptr<MT_t> MT; // the big MTasker
 std::unique_ptr<MemRecursorCache> g_recCache;
 std::unique_ptr<NegCache> g_negCache;
 
+static thread_local pdns::TaskQueue t_taskQueue;
```

Comment:
Out of curiosity, did you consider making the queue global? I know we should not push the same entry twice into different queues since we mark them as "submitted" (although, I guess we could trigger once from the PC and once from the RC via different threads?) Of course this is not easy to shard if we want to keep global ordering, but we only have relative ordering (per thread) now so I was wondering if you had another reason? To avoid contention, I guess?

Reply:
I started with a global taskQueue, but then decided to run a task at the end of a resolve always, to avoid being forced to have a thread dedicated to tasks. This meant that to avoid contention, a per-thread queue was preferred. There is indeed a scenario where the same task could end up in two queues. If one of the queue is executed first, the second task will be a no-op. If there are executed simultaneously, I guess the would be handles the same as two client-queries for the same name, i.e. chaining will take place.

Reply:
Right, that makes sense to me. Thanks!